bundles/org.openhab.binding.teslapowerwall/src/main/java/org/openhab/binding/teslapowerwall/internal/TeslaPowerwallWebTargets.java

```diff
@@ -0,0 +1,138 @@
+        return SystemStatus.parse(response);
+    }
+
+    public MeterAggregates getMeterAggregates(String email, String password)
+            throws TeslaPowerwallCommunicationException {
+        String response = invoke(getMeterAggregatesUri, email, password);
+        logger.trace("getMeterAggregates response = {}", response);
+        return MeterAggregates.parse(response);
+    }
+
+    public Operations getOperations(String email, String password) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getOperationUri, email, password);
+        logger.trace("getOperations response = {}", response);
+        return Operations.parse(response);
+    }
+
+    public String getToken(String email, String password) throws TeslaPowerwallCommunicationException {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty("username", "customer");
+        jsonObject.addProperty("password", password);
+        jsonObject.addProperty("email", email);
+        jsonObject.addProperty("force_sm_off", false);
+        logger.debug("logonjson = {}", jsonObject.toString());
+        String response = invoke(getTokenUri, "POST", "Content-Type", "application/json", jsonObject.toString());
+        JsonObject jsonResponse = JsonParser.parseString(response).getAsJsonObject();
+        String token = jsonResponse.get("token").getAsString();
+        logger.debug("Token: {}", token);
+        return token;
+    }
+
+    private String invoke(String uri, String email, String password) throws TeslaPowerwallCommunicationException {
+        if (token.isEmpty()) {
+            token = getToken(email, password);
+        }
+        return invoke(uri, "GET", "Authorization", "Bearer " + token, "");
+    }
+
+    private String invoke(String uri, String request, String contenttype, String contenttypeparam, String params)
+            throws TeslaPowerwallCommunicationException {
+        logger.debug("Calling url: {}", uri);
+        Properties headers = new Properties();
+        headers.setProperty(contenttype, contenttypeparam);
+        ByteArrayInputStream input = new ByteArrayInputStream(params.getBytes(StandardCharsets.UTF_8));
+        String response;
+        synchronized (this) {
+            try {
+                response = HttpUtil.executeUrl(request, uri, headers, input, contenttype, TIMEOUT_MS);
+            } catch (IOException ex) {
+                logger.debug("{}", ex.getLocalizedMessage(), ex);
+                // Response will also be set to null if parsing in executeUrl fails so we use null here to make the
+                // error check below consistent.
+                response = null;
+            }
+        }
+
+        if (response == null) {
+            throw new TeslaPowerwallCommunicationException(
+                    String.format("Tesla Powerwall returned error while invoking %s", uri));
+        }
+        return response;
```

Comment:
Not sure if you need to check if response is null or not. But adding a check if the response is 401/403 and throw `TeslaPowerwallAuhtenticationException` would be usefull.